src/main/java/net/openhft/chronicle/bytes/AbstractBytes.java

```diff
@@ -1085,8 +1086,9 @@ public int read(@NotNull byte[] bytes, int off, int len) throws BufferUnderflowE
     }
 
     @Override
-    public long read(long offsetInRDI, byte[] bytes, int offset, int length) throws IllegalStateException {
-        return bytesStore.read(offsetInRDI, bytes, offset, length);
+    public long read(@NonNegative long offsetInRDI, byte[] bytes, @NonNegative int offset, @NonNegative int length) throws IllegalStateException {
+        int len = Math.min(length, Maths.toUInt31(readLimit() - offsetInRDI));
```

Comment:
these commits are already in 2.22 - do you need to rebase?

Reply:
I made a branch with your test and a backport of @glukos fix for this - https://github.com/OpenHFT/Chronicle-Bytes/pull/456